File: CPP_Solutions/Longest Common Subsequence/Print_LCS.cpp
#include<iostream>
#include<algorithm>
using namespace std; 

string printLCS(char x[], char y[], int n, int m) {
    int t[n+1][m+1];
    
    for(int i = 0; i < n + 1; i++)
        t[i][0] = 0;
    for(int j = 0; j < m + 1; j++)
        t[0][j] = 0;
        
    for(int i = 1; i < n + 1; i++)
        for(int j = 1; j < m + 1; j++)
            t[i][j] = (x[i-1] == y[j-1]) ? (1 + t[i-1][j-1]) : std::max(t[i-1][j], t[i][j-1]);
            
    string lcs;
    int i = n, j= m;
    while(i > 0 && j > 0){
        if(x[i-1] == y[j-1]){
            lcs.push_back(x[i-1]);
            i--;
            j--;
        }
        else if(t[i-1][j] > t[i][j-1])
            i --;
        else
            j --;
    }
    reverse(lcs.begin(), lcs.end());
    return lcs;
}

int main() 
{ 
    char x[] = "abcdgh";
    char y[] ="abedfhr";
    int n = (sizeof(x) / sizeof(x[0])) - 1;
    int m = (sizeof(y) / sizeof(y[0])) - 1;
    string lcs = printLCS(x,y,n,m);
    cout << lcs;
    
    return 0;
}

File: Python_Solutions/Longest_Common_Subsequence/Print_LCS.py
def printLCS(x,y,n,m):
    t = [[0 for j in range(m+1)] for i in range(n+1)] # Initialise matrix with zero

    for i in range(1, n+1):
        for j in range(1, m+1):
            t[i][j] = (1 + t[i-1][j-1]) if(x[i-1] == y[j-1]) else (max(t[i-1][j], t[i][j-1]))

    index = t[n][m]
    lcs = [""] * (index) #creating a character array of length index
    i=n
    j=m
    while i > 0 and j > 0:
        if(x[i-1] == y[j-1]):
            lcs[index - 1] = x[i-1]
            i-=1
            j-=1
            index-=1
        elif(t[i-1][j] > t[i][j-1]):
            i-=1
        else:
            j-=1
    return "".join(lcs) # since lcs stores the each character as elements of an array/list

x = "abcdgh"
y = "abedfhr"
print(printLCS(x,y,len(x), len(y)))